File: src/main/java/com/gandan/practice/rx/RxUsing.java
package com.gandan.practice.rx;

import com.gandan.practice.task.ITask;
import com.sun.glass.ui.SystemClipboard;
import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func0;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * Looking for Using operation in RxJava
 * Created by chandra on 12/5/15.
 */
public class RxUsing implements ITask {

    @Override
    public void execute() {
        Observable.using(() -> {
                    System.out.println("Func 0");
                    System.out.println(">> Thread id " + Thread.currentThread().getId());
                    return "Func 0";
                }, s -> {
                    System.out.println("Func 1");
                    System.out.println(">> Thread id " + Thread.currentThread().getId());
                    return Observable.just(s).observeOn(Schedulers.newThread());
                }, s -> {
                    System.out.println("Action 1");
                    System.out.println(">> Thread id " + Thread.currentThread().getId());
                    System.out.println(s);
                }).subscribe(str -> {
                    System.out.println("String " + str);
                    System.out.println(">> Thread id " + Thread.currentThread().getId());
                }
        );
    }
}

File: src/main/kotlin/com/gandan/practice/rx/RxUsing.kt
package com.gandan.practice.rx

import rx.Observable
import rx.schedulers.Schedulers

fun main(args: Array<String>) {
    RxUsing().execute()
}

class RxUsing {

     fun execute() {
        Observable.using<Any, Any>({
            System.out.println("resource factory")
            System.out.println(">> Thread id " + Thread.currentThread().id)
            "Func 0"
        }, { s ->
            System.out.println("observable factory")
            System.out.println(">> Thread id " + Thread.currentThread().id)
            Observable.just(s).observeOn(Schedulers.newThread())
        }) { s ->
            System.out.println("dispose")
            System.out.println(">> Thread id " + Thread.currentThread().id)
            System.out.println(s)
        }.subscribe { str ->
            System.out.println("Subscribe $str")
            System.out.println(">> Thread id " + Thread.currentThread().id)
        }
    }
}
